src/tutorials/ex3.c

```diff
@@ -0,0 +1,186 @@
+  h = 1./(n-1);
+  ierr = MatCreate(PETSC_COMM_WORLD,&A);CHKERRQ(ierr);
+  ierr = MatSetSizes(A,PETSC_DECIDE,PETSC_DECIDE,n,n);CHKERRQ(ierr);
+  ierr = MatSetFromOptions(A);CHKERRQ(ierr);
+  ierr = MatSetUp(A);CHKERRQ(ierr);
+  ierr = MatGetOwnershipRange(A,&rstart,&rend);CHKERRQ(ierr);
+  ierr = MatCreateVecs(A,&x,&b);CHKERRQ(ierr);
+  ierr = VecDuplicate(x,&c);CHKERRQ(ierr);
+
+  if (!rstart) {
+    rstart = 1;
+    i      = 0; value[0] = 1.0;
+    ierr   = MatSetValues(A,1,&i,1,&i,value,INSERT_VALUES);CHKERRQ(ierr);
+    ierr   = VecSetValue(b,i,0,INSERT_VALUES);CHKERRQ(ierr);
+  }
+  if (rend == n) {
+    rend = n-1;
+    i    = n-1; value[0] = 1.0;
+    ierr = MatSetValues(A,1,&i,1,&i,value,INSERT_VALUES);CHKERRQ(ierr);
+    ierr = VecSetValue(b,i,0,INSERT_VALUES);CHKERRQ(ierr);
+  }
+  value[0] = -1.0; value[1] = 2.0; value[2] = -1.0;
+  for (i=rstart; i<rend; i++) {
+    col[0] = i-1; col[1] = i; col[2] = i+1;
+    ierr = MatSetValues(A,1,&i,3,col,value,INSERT_VALUES);CHKERRQ(ierr);
+    ierr = VecSetValue(b,i,-15*h*h*2,INSERT_VALUES);CHKERRQ(ierr);
+    ierr = VecSetValue(c,i,fobst(i,n),INSERT_VALUES);CHKERRQ(ierr);
+  }
+  ierr = MatAssemblyBegin(A,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
+  ierr = MatAssemblyEnd(A,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
+  ierr = VecAssemblyBegin(b);CHKERRQ(ierr);
+  ierr = VecAssemblyEnd(b);CHKERRQ(ierr);
+  ierr = VecAssemblyBegin(c);CHKERRQ(ierr);
+  ierr = VecAssemblyEnd(c);CHKERRQ(ierr);
+
+  ierr = MatCreateAIJ(PETSC_COMM_WORLD,PETSC_DECIDE,PETSC_DECIDE,n,n,1,NULL,0,NULL,&B);CHKERRQ(ierr);
+  ierr = MatAssemblyBegin(B,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
+  ierr = MatAssemblyEnd(B,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
+  ierr = MatShift(B,1.0);CHKERRQ(ierr);
+
+  /* Empty null space matrix for dualization */
+  ierr = MatCreateAIJ(PETSC_COMM_WORLD,PETSC_DECIDE,PETSC_DECIDE,n,0,0,NULL,0,NULL,&R);CHKERRQ(ierr);
+  ierr = MatAssemblyBegin(R,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
+  ierr = MatAssemblyEnd(R,MAT_FINAL_ASSEMBLY);CHKERRQ(ierr);
+
+  /* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+  * Setup QP: argmin 1/2 x'Ax -x'b s.t. c <= I*x
+  *  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
+  ierr = QPCreate(PETSC_COMM_WORLD,&qp);CHKERRQ(ierr);
+  /* Set matrix representing QP operator */
+  ierr = QPSetOperator(qp,A);CHKERRQ(ierr);
+
+  /* Set right hand side */
+  ierr = QPSetRhs(qp,b);CHKERRQ(ierr);
+  /* Set initial guess. 
+  * THIS VECTOR WILL ALSO HOLD THE SOLUTION OF QP */
+  ierr = QPSetInitialVector(qp,x);CHKERRQ(ierr);
+  /* Set inequality constraint c <= Bx in the form -c >= -Bx*/
+  ierr = VecScale(c,-1.0);CHKERRQ(ierr);
+  ierr = MatScale(B,-1.0);CHKERRQ(ierr);
+  ierr = QPSetIneq(qp,B,c);CHKERRQ(ierr);
+  /* Dualize QP */
+  ierr = QPSetOperatorNullSpace(qp,R);CHKERRQ(ierr);
+  ierr = PetscOptionsInsertString(NULL,"-qpt_dualize_B_nest_extension 0 -qpt_dualize_G_explicit 0");CHKERRQ(ierr); /* workaround for empty nullspace */
```

Comment:
@jkruzik Is #13 supposed to remove this?

Reply:
It should remove G_explicit=0 because there will be no  R.

The B_nest_extension=0 actually is not necessary here. On the other hand it maybe should not be the default?

Reply:
Yeah, it's not a good default. I'm now facing it in another branch - I will change the default to `-qpt_dualize_B_nest_extension 0` there.